src/android/OpenIDAuthStateManager.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright 2017 The AppAuth for Android Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package edu.berkeley.eecs.emission.cordova.jwtauth;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.support.annotation.AnyThread;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import net.openid.appauth.AuthState;
+import net.openid.appauth.AuthorizationException;
+import net.openid.appauth.AuthorizationResponse;
+import net.openid.appauth.RegistrationResponse;
+import net.openid.appauth.TokenResponse;
+
+import org.json.JSONException;
+
+import java.lang.ref.WeakReference;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An example persistence mechanism for an {@link AuthState} instance.
+ * This stores the instance in a shared preferences file, and provides thread-safe access and
+ * mutation.
+ */
+public class OpenIDAuthStateManager {
```

Comment:
I am not going to do an in-depth review of this code since it is largely unchanged from sample code on the internet. If there are specific parts you want me to take a look at, mark them, and I can do so.

Reply:
let's leave it for now. I didn't encounter any problem up to this point.